compiler/x/codegen/BinaryEvaluator.cpp

```diff
@@ -1933,28 +1933,90 @@ TR::Register *OMR::X86::TreeEvaluator::integerDualMulEvaluator(TR::Node *node, T
 
 TR::Register *OMR::X86::TreeEvaluator::integerMulEvaluator(TR::Node *node, TR::CodeGenerator *cg)
    {
-   TR::Register *targetRegister = 0;
-   TR::Node     *firstChild     = node->getFirstChild();
-   TR::Node     *secondChild    = node->getSecondChild();
-   bool          nodeIs64Bit    = TR::TreeEvaluator::getNodeIs64Bit(node, cg);
-   intptrj_t    value;
+   TR::Register *targetRegister  = 0;
+   TR::Node     *firstChild      = node->getFirstChild();
+   TR::Node     *secondChild     = node->getSecondChild();
+   bool          nodeIs64Bit     = TR::TreeEvaluator::getNodeIs64Bit(node, cg);
+   bool          isLoadConst     = secondChild->getOpCode().isLoadConst();
+   bool          isByteOperation = false;
+   bool          isMemRef        = false;
+   TR_X86OpCodes opCode          = BADIA32Op; 
+   TR_X86OpCodes regOpCode       = BADIA32Op; 
+   TR_X86OpCodes memOpCode       = BADIA32Op; 
+   TR_X86OpCodes movOpCode       = BADIA32Op;
+   int64_t       value;
 
    if (node->isDualCyclic())
       {
       return TR::TreeEvaluator::integerDualMulEvaluator(node, cg);
       }
 
-   if (secondChild->getOpCode().isLoadConst() &&
-      (value = TR::TreeEvaluator::integerConstNodeValue(secondChild, cg)))
+   // Opcode evaluation for bmul, smul imul and lmul
+   switch (secondChild->getDataType())
+      {
+      case TR::Int8:
+         isMemRef = secondChild->getOpCode().isMemoryReference(); 
+         opCode = isMemRef ? IMUL1AccMem : IMUL1AccReg;
+     
+         regOpCode = IMUL1AccReg;
+         memOpCode = IMUL1AccMem;
+         movOpCode = MOV1RegReg;
+
+         isByteOperation = true;
+         break;
+
+      case TR::Int16:
+         if(isLoadConst)
+            {
+            value = secondChild->getShortInt();
+            isMemRef = firstChild->getOpCode().isMemoryReference(); 
+            if (isMemRef)
```

Comment:
I think there is bug here with all the `isMemRef`. Whether we can use the memory reference version of the instruction also depends on whether the first child has multiple uses. With the current code we can end up calling `generateRegRegImmInstruction` with a `IMUL2RegMemImms` or `IMULRegMemImms(nodeIs64Bit)` when `firstChild->getReferenceCount() > 1`

Reply:
I think, assigning those values upfront makes the code hard to follow and easy to introduce bug since you need to repeat the if else logic below. I would suggest replacing uses of the opcode during the if else logic into queries like `getMulOpForInMemoryUpdate(Node*node)`